Template/Trigger/ElementVisibilityTrigger.web.js

```diff
@@ -235,9 +230,82 @@
+                    root: null, // document's viewport as the container.
+                    rootMargin: '0px',
+                    threshold: (minPercentVisible / 100)
+                };
+                var observerIntersection = new IntersectionObserver(function (entries) {
+                    interSectionCallback(entries, triggerEvent, observerIntersection);
+                }, interSectionObserverOptions);
+
+                if (selectors) {
+                    TagManager.dom.bySelector(selectors).forEach(function (element) {
+                        observerIntersection.observe(element);
+                    });
+                }
+            }
+
+        }
+
+        function interSectionCallback(entries, triggerEvent, observer) {
+            var dom = TagManager.dom;
+            entries.forEach(function (entry) {
+                if (entry.intersectionRatio > 0) {
+                    if (blockTrigger || (onlyOncePerElement && isNodeEventTriggered(entry.target))) {
+                        return;
+                    }
+                    triggerEvent({
+                        event: 'mtm.ElementVisibility',
+                        'mtm.elementVisibilityPercentage': minPercentVisible,
+                        'mtm.elementVisibilityId': dom.getElementAttribute(entry.target, 'id'),
+                        'mtm.elementVisibilityClasses': dom.getElementClassNames(entry.target),
+                        'mtm.elementVisibilityText': TagManager.utils.trim(entry.target.innerText),
+                        'mtm.elementVisibilityNodeName': entry.target.nodeName,
+                        'mtm.elementVisibilityUrl': entry.target.href || dom.getElementAttribute(entry.target, 'href'),
+                    });
+
+                    if (fireTriggerWhen === 'oncePage') {
+                        blockTrigger = true;
+                        TagManager.dom.bySelector(selectors).forEach(function (element) {
+                            observer.unobserve(element);
+                            triggeredNodes.push(element);
+                        });
+                    } else if(onlyOncePerElement) {
+                        observer.unobserve(entry.target);
+                        triggeredNodes.push(entry.target); // to avoid possible memory leaks as much as possible we add onceElement only when needed
+                    }
+                }
+            });
+        }
+
+        function isNodeEventTriggered(node) {
+            for (var j = 0; j < triggeredNodes.length; j++) {
+                if (node === triggeredNodes[j]) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
         this.setUp = function (triggerEvent) {
             this.scrollIndex = TagManager.window.onScroll(checkVisiblity(triggerEvent));
             TagManager.dom.onLoad(checkVisiblity(triggerEvent));
+            setIntersectionObserver(triggerEvent);
```

Comment:
I'm thinking we shouldn't load the intersection observer at this point. If the page is not fully loaded yet, then some elements may be visible for just a tiny short time before the CSS is applied and the element pushed down maybe.

Maybe we could instead try to run this on "onLoad" like `TagManager.dom.onLoad(setIntersectionObserver(triggerEvent));`?

would intersection observer let us know if it was visible previously before onload? (just in case the onload event is quite delayed)

Reply:
or I think `checkVisiblity` detects elements that may have been visible before "onload" by checking the scrolling position and assumes any element above this would have been visible at some stage. Not sure.

In that case could also execute something like 

```
TagManager.dom.onLoad(checkVisiblity(triggerEvent)); // detect any elements that have been visible previously
TagManager.dom.onLoad(setIntersectionObserver(triggerEvent)); // detect any matching element when it becomes visible
```

Reply:
Moved to dom onload as suggested above